HW4/HW4.py

```diff
@@ -0,0 +1,381 @@
+        p_value = "{:.6e}".format(linear_regression_model_fitted.pvalues[1])
+        # print(f"t value: {t_value}", f"p value: {p_value}")
+        # Plot the figure
+        fig = px.scatter(x=df[x], y=df[y], trendline="ols")
+        fig.update_layout(
+            title=f"Variable: {x}: (t-value={t_value}) (p-value={p_value})",
+            xaxis_title=f"Variable: {x}",
+            yaxis_title="y",
+        )
+        filename = x + " ts and ps.html"
+        fig.write_html(file=filename, include_plotlyjs="cdn")
+        return filename, t_value, p_value
+
+    list_of_files = []
+    # for loop here
+    for predictor in predictors:
+        predictor_type = predictor_types.get(predictor)
+        x = predictor
+        # generate plots to inspect each predictor
+        # plot code was generously donated by dafrenchyman
+        if response_type in ["boolean"]:
+            if predictor_type in ["boolean", "categorical"]:
+                # Heatplot here
+                file_heatplot = Heatplot(x)
+                list_of_files.append(file_heatplot)
+
+            elif predictor_type == "continuous":
+                # violin here
+                file_violin = Violin(x)
+                list_of_files.append(file_violin)
+
+                # distribution here
+                file_distribution = Distribution(x)
+                list_of_files.append(file_distribution)
+
+        elif response_type in ["continuous"]:
+            if predictor_type in ["boolean", "categorical"]:
+                # second violin here
+                # distribution2 here
+                file_violin2, file_distribution2 = ViolinAndDist(x)
+                list_of_files.append(file_violin2)
+                list_of_files.append(file_distribution2)
+
+            elif predictor_type == "continuous":
+                # scatter here
+                file_scatter = Scatter(x)
+        # calculate the ranking algorithms
+        if predictor_type == "continuous":
+            if response_type == "continuous":
+                # regression here
+                file_regression, t_value, p_value = Regression(x)
+                list_of_files.append(file_regression)
+
+            elif response_type == "boolean":
+                # logistic regression: boolean response
+                logistic_regression_model = statsmodels.api.Logit(df[y], df[x]).fit()
+                print(logistic_regression_model.summary())
+
+                # I had to remove this to accomodate flake8. I'm not happy about that
+                # try:
+                # t_value = round(logistic_regression_model.tvalues[1], 6)
+                # p_value = "{:.6e}".format(logistic_regression_model.pvalues[1])
+                # except:
+                # pass
+
+    # difference with mean of response along with its plot (weighted and unweighted)
+    # mean of response per bin
+    # 1/number of bins * sum(each bin mean - pop mean)^2
+    def avg_mean_of_response(predictor):
```

Comment:
this function is very confusing but I am also very burnt out while reviewing this so its hard to examine